domains/transfer/Transfer.API/API/Transfer/Api/Controllers/TransferAgreementProposalController.cs

```diff
@@ -0,0 +1,138 @@
+using API.Transfer.Api.Repository;
+using FluentValidation;
+using FluentValidation.AspNetCore;
+
+namespace API.Transfer.Api.Controllers;
+
+[Authorize]
+[ApiController]
+[Route("api/transfer-agreement-proposals")]
+public class TransferAgreementProposalController : ControllerBase
+{
+    private readonly ITransferAgreementProposalRepository repository;
+    private readonly IValidator<CreateTransferAgreementProposal> createTransferAgreementProposalValidator;
+    private readonly ITransferAgreementRepository transferAgreementRepository;
+
+    public TransferAgreementProposalController(ITransferAgreementProposalRepository repository,
+        IValidator<CreateTransferAgreementProposal> createTransferAgreementProposalValidator,
+        ITransferAgreementRepository transferAgreementRepository)
+    {
+        this.repository = repository;
+        this.createTransferAgreementProposalValidator = createTransferAgreementProposalValidator;
+        this.transferAgreementRepository = transferAgreementRepository;
+    }
+
+    /// <summary>
+    /// Create TransferAgreementProposal
+    /// </summary>
+    /// <param name="request">The request object containing the StartDate, EndDate and ReceiverTin needed for creating the Transfer Agreement.</param>
+    /// <response code="201">Created</response>
+    /// <response code="409">There is already a Transfer Agreement with this company tin within the selected date range</response>
+    [ProducesResponseType(typeof(Guid), 201)]
+    [ProducesResponseType(typeof(string), 409)]
+    [HttpPost]
+    public async Task<ActionResult> CreateTransferAgreementProposal(CreateTransferAgreementProposal request)
+    {
+        var companySenderId = Guid.Parse(User.FindSubjectGuidClaim());
+        var companySenderTin = User.FindSubjectTinClaim();
+        var companySenderName = User.FindSubjectNameClaim();
+
+        var validateResult = await createTransferAgreementProposalValidator.ValidateAsync(request);
+        if (!validateResult.IsValid)
+        {
+            validateResult.AddToModelState(ModelState);
+            return ValidationProblem(ModelState);
+        }
+
+        var newProposal = new TransferAgreementProposal
+        {
+            SenderCompanyId = companySenderId,
+            SenderCompanyTin = companySenderTin,
+            SenderCompanyName = companySenderName,
+            Id = Guid.NewGuid(),
+            ReceiverCompanyTin = request.ReceiverTin,
+            StartDate = DateTimeOffset.FromUnixTimeSeconds(request.StartDate),
+            EndDate = request.EndDate == null ? null : DateTimeOffset.FromUnixTimeSeconds(request.EndDate.Value)
+        };
+
+        var hasConflict = await transferAgreementRepository.HasDateOverlap(newProposal);
+
+        if (hasConflict)
+        {
+            return Conflict("There is already a Transfer Agreement with this company tin within the selected date range");
+        }
+
+        await repository.AddTransferAgreementProposal(newProposal);
```

Comment:
Should there be a try catch to ensure no concurrent request?

Reply:
Concurrent requests can occur but is not that important since this is only proposals and they are deleted when they are older than 14 days